### PR TITLE
Text, markdown and code editor

### PR DESCRIPTION
**Pull request by ianopolous**
This implements a text, code and markdown editor

Important points:
1) I've made save only work on request from outside the component
2) This made me realise we also need a markdown ***viewer***! for nice todo lists etc.
3) The save icon changes to an hourglass during saving. 
4) Errors saving are only logged to the console for now.

### CODE REVIEW

**Comment by ianopolous** [comment]
Screenshot_2019-07-22 Peergos(1)

**Comment by ianopolous** [comment]
One issue I've noticed is that it doesn't truncate the file if you make it shorter. This will need an new api from Java.

File: src/components/filesystem/index.js [hunk 1]
```diff
@@ -850,7 +851,10 @@ module.exports = {
 		var that = this;
 		this.confirmView(file, () => {that.showGallery = true;});
 	    } else if (mimeType === "text/plain") {
-		this.showTextViewer = true;
+		if (file.getName().endsWith(".md"))
```
Comment (kevodwyer) [inline src/components/filesystem/index.js]: what about other extensions?

File: src/components/code-editor/index.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	    window.addEventListener('message', function (e) {
+		// Normally, you should verify that the origin of the message's sender
+		// was the origin and source you expected. This is easily done for the
+		// unsandboxed frame. The sandboxed frame, on the other hand is more
+		// difficult. Sandboxed iframes which lack the 'allow-same-origin'
+		// header have "null" rather than a valid origin. This means you still
+		// have to be careful about accepting data via the messaging API you
+		// create. Check that source, and validate those inputs!
+		if (e.origin === "null" && e.source === iframe.contentWindow) {
+		    console.log('Message from Iframe: ' + e.data);
```
Comment (kevodwyer) [inline src/components/code-editor/index.js]: debugging code

File: assets/apps/code-editor/index.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
+    <title>Document editor</title>
```
Comment (kevodwyer) [inline assets/apps/code-editor/index.html]: Elsewhere we use Code editor. Document editor means word processor to most people.

File: src/components/filesystem/index.js (hunk 1 above)
Comment (ianopolous) [reply src/components/filesystem/index.js]: Yep. I will add some more. I mainly want plain text

File: assets/apps/code-editor/index.html (hunk 3 above)
Comment (ianopolous) [reply assets/apps/code-editor/index.html]: Yep I want to target text editing

File: src/components/filesystem/index.js (hunk 1 above)
Comment (ianopolous) [reply src/components/filesystem/index.js]: I made it open all text files.

**Comment by ianopolous** [comment]
Scala editing!
Screenshot_2019-07-23 Peergos

**Comment by ianopolous** [comment]
This is now good to go (after a submodule update)

**Pull request merged**